kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/registry/RegistryStorage.kt

```diff
@@ -0,0 +1,65 @@
+package com.kotlindiscord.kord.extensions.registry
+
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * Interface for interaction based registries like ComponentRegistry and ApplicationCommandRegistry.
+ *
+ * The purpose of this interface is to provide a generic way to store Components/ApplicationCommands
+ * in a dynamic manner.
+ */
+public interface RegistryStorage<K, T> {
+
+    /**
+     * Let's the registry know about the specified type, this may store the object in a local map,
+     * which is used for reconstructing later.
+     */
+    public suspend fun register(data: T)
+
+    /**
+     * Creates or updates an existing entry at the given unique key.
+     *
+     * This may deconstruct the given data and only persists a partial object.
+     */
+    public suspend fun set(id: K, data: T)
+
+    /**
+     * Reads a value from the registry at the given key.
+     *
+     * This may reconstruct the data from a partial object.
+     */
+    public suspend fun get(id: K): T?
+
+    /**
+     * Deletes a value from the registry with the given key.
+     *
+     * The return value may be a reconstructed object from partial data.
+     */
+    public suspend fun remove(id: K): T?
+
+    /**
+     * Creates a flow of all entries in this registry.
+     *
+     * The objects in this flow may be reconstructed from partial data.
+     */
+    public fun entryFlow(): Flow<StorageEntry<K, T>>
+
+    /**
+     * Constructs a unique key for the given data.
+     */
+    public fun constructUniqueIdentifier(data: T): String
+
+    /**
+     * Data class to represent an entry in the [RegistryStorage].
+     */
+    public data class StorageEntry<K, V>(
```

Comment:
This is basically just a `Pair`. Any reason why you didn't use that?

Reply:
I don't like the .0 .1 syntax on `Pair` and overall this explain in which scope this "Pair" lives.
Also I was playing with putting the registry into the entry as an additional reference, yet decided against it. This would also allow for further extension of this without breaking other's code.

Reply:
Hmm, fair point.